Overhaul incremental compiler

oehme opened the pull request:
This started as an effort to fix #4951, but quickly turned into a total overhaul of the incremental compiler code, which reduced its complexity and also fixed a few other issues. 

Most notably, api/implementation separation wasn't working for classpath changes (#13499), so Gradle was recompiling way too much in multi-module projects. I've added a test to make sure this doesn't break again.

The main change is that the incremental compiler no longer uses absolute paths to identify classpath entries and whether they've changed or been reordered. Instead, the ClassSetAnalysis of all classpath entries is merged into one big analysis, which is then diffed against the analysis of the last build. This analysis is then stored in the build cache, along with the ClassSetAnalysis of the output directory. The serialized form of ClassSetAnalysis is now space-optimized, so that this additional build cache entry remains small enough to be worth the trade-off.

I've also changed the messages that Gradle provides when it does a full recompilation. It is now explicit about which change caused the full recompilation. This should make it easier to understand and debug.

Handling of newly added package-info and module-info files was broken since the introduction of the `SourceFileClassNameConverter`. This is fixed and tested now too.

I've also added some JavaDoc to the most important classes in the incremental compiler and renamed a few of their methods for clarity.

While trying to performance-test this, I had trouble generating a large enough dependency graph, so I also fixed the performance of our test project generator. In the end, it still wasn't a great test, since it creates a very shallow tree. So instead I tested this on `gradle/gradle` itself. Doing an ABI change to `GUtil`, 7.0-milestone-3 takes **~105s** while this branch only takes **~52s** thanks to the fixed classpath change handling. The CPU overhead of the incremental compiler is roughly the same as before and it is now cacheable. The total size of all added analysis outputs is about 24MB or 125KB per subproject. The daemon memory footprint is now smaller, because we only keep the ABI of classpath entries and discard any information about implementation details, since these can't affect downstream classes.

Fixes #4951
Fixes #14917
Fixes #13499
Fixes #9790
Fixes #9052

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

wolfs commented:
We have the feature for detecting if something has been added in the middle of the classpath provided by `InputChanges`. Would it be feasible to use this information for what things changed instead trying to diff the two classpaths again?

oehme commented:
No, because `FileChange` does not tell me what the hash of the old file was, so I can't look up its class analysis.

oehme commented:
> though it should be possible to somehow go from FileChange to the corresponding ClasspathEntrySnapshot, right

The classpath snapshots are identified by hash (since they go in the build cache) and the `FileChange` API does not tell me the hashes. I would need both the old and new hash of the changed file.

Also, the API would have to distinguish between "content changed" and "order changed". The former can often be done incrementally, the latter can lead to class shadowing and requires a full recompile.

wolfs commented:
> Also, the API would have to distinguish between "content changed" and "order changed". The former can often be done incrementally, the latter can lead to class shadowing and requires a full recompile.

The API already distinguishes between "Content changed" and "Order changed":
- Content changed: `ChangeType.MODIFIED`
- Order changed: `ChangeType.{ADDED,REMOVED}`

wolfs commented:
I am not proposing to replace the snapshotting by `InputChanges`, but to replace the diff calculation here by using the data from `InputChanges`: https://github.com/gradle/gradle/blob/bf41177eaeabf70cdff385cf10d8d3226a24fc9c/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/AbstractRecompilationSpecProvider.java#L63-L96

oehme commented:
Just discussed via Slack and the above code no longer exists. We could probably use the InputChanges to detect some non-incremental scenarios more quickly, but overall it would not make the code in `CurrentCompilation` any simpler. I'm gonna go ahead and simplify it a little more by always doing a full recompile when the length of the classpath lists is different. The likelihood of getting useful incrementality in that case is small enough that I don't think having that special case is worth it.

oehme commented on subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractSourceIncrementalCompilationIntegrationTest.groovy:
```diff
@@ -1099,26 +1099,29 @@ sourceSets {
+        executer.stop()
+        file("user-home").deleteDir()
+
+        and: "We're no longer using the old dependency"
```
Otherwise it's in the cache because we analyse it for the CurrentCompilation

oehme commented on subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractSourceIncrementalCompilationIntegrationTest.groovy:
```diff
@@ -1099,26 +1099,29 @@ sourceSets {
+        file("user-home").deleteDir()
+
+        and: "We're no longer using the old dependency"
+        buildFile << "dependencies { implementation 'org.apache.commons:commons-lang3:3.9' }"
```
Using this specific dependency, because it has a different hash, but no changes to constants or other full recompile reasons

oehme commented on subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalResultStoringCompiler.java:
```diff
@@ -89,27 +86,13 @@ private AnnotationProcessingData getAnnotationProcessingResult(JavaCompileSpec s
-        return new AnnotationProcessingData(intern(generatedTypesByOrigin), intern(aggregatedTypes), intern(aggregatingTypes), generatedResourcesByOrigin, aggregatingResources, processingResult.getFullRebuildCause());
-    }
-
-    private Set<String> intern(Set<String> types) {
```
We no longer keep this in an in-memory cache, but on disk, so interning is not necessary.

oehme commented on subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/BuildCacheClasspathEntrySnapshotCache.java:
```diff
@@ -0,0 +1,163 @@
+/**
+ * Stores classpath entry snapshots in the build cache and loads them from there when they can't be found in the given delegate.
+ */
+public class BuildCacheClasspathEntrySnapshotCache implements ClasspathEntrySnapshotCache {
```
@wolfs I'd like your feedback on this class - This pattern could be generalized into a build cache decorator for any kind of expensive data. For example, we could also use this pattern to store the Classpath/CompileClasspath snapshots for external dependencies, so we don't have to recompute them every time (e.g. on ephemeral CI nodes this can cost a lot of time).

melix replied:
You should probably use the new `ExecutionEngine` instead. The idea would be to create a `UnitOfWork` to perform the class analysis. The engine should handle storing in the cache for you, so you don't need to implement your own cache keys for example.

oehme commented:
Going back to draft, since I have some more ideas how to make this simpler and even restore the incremental corner cases mentioned above.

lptr commented:
> > You should probably use the new ExecutionEngine instead. The idea would be to create a UnitOfWork to perform the class analysis. The engine should handle storing in the cache for you, so you don't need to implement your own cache keys for example.
> 
> I looked at it, but it seems overblown for this use case. I just want to store a tiny piece of data in the cache, not create output files etc.

Take a look at how `UnitOfWork` is implemented for generating project accessors here: https://github.com/gradle/gradle/blob/0f61831dba98d28b62bbaacc06516e2d0c6d4f60/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/AccessorsClassPath.kt#L103-L166

The idea is that this is how you encapsulate any kind of work that needs to be executed as part of the build going forward. It is supposed to be lightweight and simple enough to implement for even small units of work; if you find something that seems overcomplicated as it is now, or some feature is missing for your use case, we can always discuss how to improve things. We should not introduce new caches if we can avoid it, as well as talking directly to the build cache.

Happy to discuss further with you if you want.

oehme commented on subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java:
```diff
@@ -177,6 +178,19 @@ protected File getSourceClassesMappingFile() {
+     * @since 7.1
+     */
+    @OutputFile
+    protected File getPreviousCompilationData() {
```
This is auto-computed and not user-configurable, so I don't think we gain anything from using a Property here.

oehme commented:
I've completely reworked this 
1. The classpath diff is now done by merging all the entries and then comparing the merged result. This unlocked a ton of other simplifications and eventually made me aware that we didn't have abi/impl separation for classpath changes! I've fixed that by making the classpath analysis use the exact same code path we use for the normal class analysis.
2. I removed the need for BuildCache or ExecutionEngine. Instead, all the necessary state is stored in `PreviousCompilationData`. I've made some optimizations so that the data is small enough to make this feasible. This is both simpler and also a better example for other plugin authors looking to implement something similar for their language or tool.

I've tried this branch out on gradle/gradle, making an ABI change to GUtil.java. The results are very exciting: It takes **107s** on 7.0-milestone-3, but only **52s** on this branch. The sum of all analysis files is 25MB (about 100KB per compile task).

oehme commented on subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/RelevantProjectsRegistry.kt:
```diff
@@ -31,7 +31,7 @@ class RelevantProjectsRegistry : ProjectAccessHandler {
     fun relevantProjects(nodes: List<Node>): Set<ProjectInternal> =
         targetProjects +
             nodes.mapNotNullTo(mutableListOf()) { node ->
-                node.owningProject?.takeIf { it.parent != null }
```
@eskatos was there a reason for this? All tests pass and it fixes a bug where the build directory of the root project would not be restored.

oehme commented:
The small regression for the largeMonolithicProject is expected - We are storing more state and that project doesn't benefit from it at all, because everything is in a single module. What's the process for accepting such a change from a branch? Do I have to ignore it until it is merged and a new nightly is published?

oehme commented:
@melix I think this is ready for a serious look from you now. I've updated the title and description to match the scope of the changes. It's probably best to look at them as a whole now, since I've gone back and forth a lot in the individual commits.

melix commented:
Excellent, thanks, I'll take a look!

melix commented on subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java:
```diff
@@ -282,19 +282,31 @@ private JavaToolchainSpec determineExplicitToolchain() {
         if (sourceClassesMappingFile == null) {
-            File tmpDir = getServices().get(TemporaryFileProvider.class).newTemporaryFile(getName());
-            sourceClassesMappingFile = new File(tmpDir, "source-classes-mapping.txt");
+            sourceClassesMappingFile = new File(getTemporaryDir(), "source-classes-mapping.txt");
```
:o: I wonder if we should move the file _out_ of the temporary directory, now that it's a real output.

oehme replied:
I think it's fine like this as long as we consider this an internal detail of the task. If we ever make this configurable or let users read our analysis file (effectively making it API), then I'd agree putting it in a more prominent, configurable location.

oehme replied:
Note that the "temp" dir is still in the build folder, not in `/tmp`. So it's not temporary at all :)

melix commented on subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java:
```diff
@@ -282,19 +282,31 @@ private JavaToolchainSpec determineExplicitToolchain() {
+    @OutputFile
+    protected File getPreviousCompilationData() {
+        if (previousCompilationDataFile == null) {
+            previousCompilationDataFile = new File(getTemporaryDir(), "previous-compilation-data.bin");
```
:o: Same here, I think we should probably move it out of the temp dir.

melix reviewed (approved):
Thanks a lot for this PR, @oehme ! I looked into it and didn't see anything obviously wrong. I checked out the branch locally, make some experiments and everything looks good!

I left a few minor comments, let me know if you plan to address them or not. I could take over polishing.

oehme commented on subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractSourceIncrementalCompilationIntegrationTest.groovy:
```diff
@@ -1079,8 +1079,15 @@ sourceSets {
+        """
         def packageFile = file("src/main/${languageName}/foo/package-info.${languageName}")
-        packageFile.text = """package foo;"""
+        packageFile.text = """@Deprecated package foo;"""
```
Yes, but then the test is the same as "adding a package-info", because a package-info file without any annotations is not compiled into a class file at all.

oehme commented on subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassDependentsAccumulator.java:
```diff
@@ -29,39 +29,39 @@
 
 public class ClassDependentsAccumulator {
```
Yes, but that's such a minor issue that I don't think it's worth complicating the code for. The cost of analysis is in actually looking at the classes, not in aggregating the results. We could stop looking at classes as soon as there is a full rebuild cause. But even that is probably not important, because broken jars should be a rare exception.

oehme commented on subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java:
```diff
@@ -66,16 +74,46 @@ public ClassSetAnalysis(ClassSetAnalysisData classAnalysis, AnnotationProcessing
+    /**
+     * An efficient version of {@link #getTransitiveDependents(String, IntSet)} when several classes have changed.
+     */
+    public DependentsSet getTransitiveDependents(Iterable<String> classes, Map<String, IntSet> constants) {
```
I'll do that because I think it adds a little clarity (that we expect a finite collection), but in my profiling sessions I never ever saw this code taking any time at all :)

oehme commented on subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java:
```diff
@@ -66,16 +74,46 @@ public ClassSetAnalysis(ClassSetAnalysisData classAnalysis, AnnotationProcessing
+        return new ClassSetDiff(allChanges, changedConstants);
+    }
+
+    private Map<String, IntSet> getChangedConstants(ClassSetAnalysis other, DependentsSet affectedClasses) {
```
Sure, no problem, I'll align that with your conventions.

oehme commented on subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/DependentsSet.java:
```diff
@@ -50,6 +52,31 @@ public static DependentsSet empty() {
         return EmptyDependentsSet.INSTANCE;
     }
 
+    public static DependentsSet merge(Collection<DependentsSet> sets) {
```
Will do.

oehme commented on subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/CurrentCompilationAccess.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+
+    public ClassSetAnalysisData getClasspathSnapshot(final Iterable<File> entries) {
```
We don't have to use intermediate Lists, the snapshotter could create a Stream out of the Iterable if we wanted to. Also I find the Stream API to be very limiting, e.g. not being able to know the size (even though we do know), whether it's empty etc and only being able to iterate over it once. I think Iterable or Collection are a better default for things that we hold in memory.

melix reviewed (approved):
LGTM!

melix commented:
@bot-gradle test and merge

bot-gradle commented:
Error while scheduling pre-tested commit, please contact the administrator.

melix commented:
@bot-gradle test and merge

bot-gradle commented:
Error while scheduling pre-tested commit, please contact the administrator.

blindpirate commented:
@bot-gradle test and merge

bot-gradle commented:
Error while scheduling pre-tested commit, please contact the administrator.

blindpirate commented:
@bot-gradle test and merge

bot-gradle commented:
Error while scheduling pre-tested commit, please contact the administrator.

blindpirate commented:
@bot-gradle test and merge

blindpirate commented:
@bot-gradle test and merge

bot-gradle commented:
Error while scheduling pre-tested commit, please contact the administrator.

blindpirate commented:
@bot-gradle test and merge

bot-gradle commented:
Your PR is queued because #16031 is being built.

bot-gradle commented:
OK, I've already triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=42411852) for you.

bot-gradle commented:
[Pre-tested commit build](https://builds.gradle.org/viewLog.html?buildTypeId=Gradle_Master_Check_Stage_ReadyforNightly_Trigger&buildId=42411852) failed.

melix commented:
@bot-gradle test and merge

bot-gradle commented:
Your PR is queued because #16713 is being built.

bot-gradle commented:
OK, I've already triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=42415905) for you.

bot-gradle commented:
[Pre-tested commit build](https://builds.gradle.org/viewLog.html?buildTypeId=Gradle_Master_Check_Stage_ReadyforNightly_Trigger&buildId=42415905) failed.

blindpirate commented:
@bot-gradle test and merge

bot-gradle commented:
OK, I've already triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=42421361) for you.

bot-gradle commented:
[Pre-tested commit build](https://builds.gradle.org/viewLog.html?buildTypeId=Gradle_Master_Check_Stage_ReadyforNightly_Trigger&buildId=42421361) failed.

blindpirate commented:
@bot-gradle test and merge

bot-gradle commented:
OK, I've already triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=42423163) for you.
